alignment_file issues

Some of these will make the tutorial harder so it would be nice if you could look at them asap:

  * `alignment_file_output` and `alignment_file_output_options` are not part of the `alignment_file` module in the documentation
  * compare_aligned_values and get_cigar_string are not in detail as planned; constraints for get_cigar_string are not guarded
  * write_range constraints are not guarded
  * sam_format and alignment_file_output protected sections are not marked as private although it is detail

on recent doxygen version:

```
include/seqan3/io/alignment_file/output.hpp:617: warning: Found ';' while parsing initializer list! (doxygen could be confused by a macro call without semicolon)
```

--- a/include/seqan3/io/alignment_file/detail.hpp
+++ b/include/seqan3/io/alignment_file/detail.hpp
@@ -43,6 +43,7 @@
 #include <sstream>
 
 #include <seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp>
+#include <seqan3/core/concept/tuple.hpp>
 #include <seqan3/std/concepts>
 #include <seqan3/std/ranges>
 
@@ -50,10 +51,8 @@
 
 namespace seqan3::detail
 {
-
 /*!\brief Compares two aligned sequence values and returns their CIGAR operation.
  * \ingroup alignment_file
- * \relatesalso seqan3::aligned_sequence_concept
  * \tparam reference_char_type Must be equality comparable to seqan3::gap.
  * \tparam query_char_type     Must be equality comparable to seqan3::gap.
  * \param  reference_char      The aligned character of the reference to compare.
@@ -93,6 +92,7 @@ namespace seqan3::detail
  * in the extended CIGAR alphabet (\p extended_cigar = `true`) the function
  * will return an 'X' since the bases are aligned but are not
  * equal.
+ * \sa seqan3::aligned_sequence_concept
  */
 template<typename reference_char_type, typename query_char_type>
 //!\cond
@@ -119,26 +119,19 @@ char compare_aligned_values(reference_char_type const reference_char,
 /*!\brief Transforms an alignment represented by two aligned sequences into the
  *        corresponding CIGAR string.
  * \ingroup alignment_file
- * \relatesalso seqan3::aligned_sequence_concept
- * \tparam ref_seq_type    Must model seqan3::aligned_sequence_concept.
- * \tparam query_seq_type  Must model seqan3::aligned_sequence_concept.
- * \param  alignment       The alignment, represented by a pair of aligned sequences,
- *                         to be transformed into CIGAR_vector based on the
- *                         second (query) sequence.
+ *
+ * \tparam ref_seq_type    Must model std::ranges::ForwardRange. The value_type must
+ *                         be equality comparable to seqan3::gap.
+ * \tparam query_seq_type  Must model std::ranges::ForwardRange. The value_type must
+ *                         be equality comparable to seqan3::gap.
+ * \param  ref_seq         The reference sequence to compare against the query sequence.
+ * \param  query_seq       The query sequence to build the CIGAR string for.
  * \param  query_start_pos The start position of the alignment in the query
  *                         sequence indicating soft-clipping.
  * \param  query_end_pos   The end position of the alignment in the query
  *                         sequence indicating soft-clipping.
  * \param  extended_cigar  Whether to print the extended cigar alphabet or not. See CIGAR operation.
- * \returns An std::string representing the alignment.
- *
- * \details
- *
- * \attention Note that CIGAR elements (respectively by their
- *            CIGAR operation) are always related to on one of the two sequences
- *            in a pairwise alignment. In this case, the resulting cigar_vector
- *            is based on sequence at the second position of the \p alignment pair,
- *            namely the query sequence.
+ * \returns An std::string representing the alignment as a CIGAR string.
  *
  * ### Theoretical Example:
  *
@@ -149,26 +142,28 @@ char compare_aligned_values(reference_char_type const reference_char,
  * |||X  |||X|  |
  * ATGCCCCGTTG--C
  * ```
- * In this case, the function seqan3::detail::compare_aligned_values will return
- * The following cigar string when printed: "4M2I5M2D1M". The extended cigar
+ * In this case, the function seqan3::detail::get_cigar_string will return
+ * the following cigar string when printed: "4M2I5M2D1M". The extended cigar
  * string would look like this: "3=1X2I3=1X1=2D1=".
+ * \sa seqan3::aligned_sequence_concept
  */
 template<std::ranges::ForwardRange ref_seq_type, std::ranges::ForwardRange query_seq_type>
-//\!cond
+//!\cond
     requires std::EqualityComparableWith<gap, value_type_t<ref_seq_type>> &&
              std::EqualityComparableWith<gap, value_type_t<query_seq_type>>
-//\!endcond
-std::string get_cigar_string(std::pair<ref_seq_type, query_seq_type> const & alignment,
-                             uint32_t query_start_pos = 0,
-                             uint32_t query_end_pos = 0,
-                             bool extended_cigar = false)
+//!\endcond
+std::string get_cigar_string(ref_seq_type && ref_seq,
+                             query_seq_type && query_seq,
+                             uint32_t const query_start_pos = 0,
+                             uint32_t const query_end_pos = 0,
+                             bool const extended_cigar = false)
 {
-    if (std::get<0>(alignment).size() != std::get<1>(alignment).size())
+    if (ref_seq.size() != query_seq.size())
         throw std::logic_error("The aligned sequences must have the same length.");
 
     std::ostringstream result;
 
-    if (!std::get<0>(alignment).size())
+    if (!ref_seq.size())
         return std::string(); // return empty string if sequences are empty
 
     // Add (S)oft-clipping at the start of the read
@@ -178,11 +173,11 @@ std::string get_cigar_string(std::pair<ref_seq_type, query_seq_type> const & ali
     // Create cigar string from alignment
     // -------------------------------------------------------------------------
     // initialize first operation:
-    char tmp_char{compare_aligned_values(std::get<0>(alignment)[0], std::get<1>(alignment)[0], extended_cigar)};
+    char tmp_char{compare_aligned_values(ref_seq[0], query_seq[0], extended_cigar)};
     unsigned tmp_length{0};
 
     // go through alignment columns
-    for (auto column : ranges::view::zip(std::get<0>(alignment), std::get<1>(alignment)))
+    for (auto column : ranges::view::zip(ref_seq, query_seq))
     {
         char next_op = compare_aligned_values(std::get<0>(column), std::get<1>(column), extended_cigar);
 
@@ -205,6 +200,59 @@ std::string get_cigar_string(std::pair<ref_seq_type, query_seq_type> const & ali
         result << query_end_pos << 'S';
 
     return result.str();
+}
+
+/*!\brief Transforms an alignment represented by two aligned sequences into the
+ *        corresponding CIGAR string.
+ * \ingroup alignment_file
+ *
+ * \tparam alignment_type  Must model the seqan3::tuple_like_concept and must
+ *                         have std::tuple_size 2. Each tuple element must model
+ *                         std::ForwardRange and its value_type must be comparable
+ *                         to seqan3::gap.
+ * \param  alignment       The alignment, represented by a pair of aligned sequences,
+ *                         to be transformed into CIGAR_vector based on the
+ *                         second (query) sequence.
+ * \param  query_start_pos The start position of the alignment in the query
+ *                         sequence indicating soft-clipping.
+ * \param  query_end_pos   The end position of the alignment in the query
+ *                         sequence indicating soft-clipping.
+ * \param  extended_cigar  Whether to print the extended cigar alphabet or not. See CIGAR operation.
+ * \returns An std::string representing the alignment as a CIGAR string.
+ *
+ * \details
+ *
+ * \attention Note that CIGAR elements (respectively by their
+ *            CIGAR operation) are always related to on one of the two sequences
+ *            in a pairwise alignment. In this case, the resulting cigar_vector
+ *            is based on sequence at the second position of the \p alignment pair,
+ *            namely the query sequence.
+ *
+ * ### Theoretical Example:
+ *
+ * The following alignment reference sequence on top and the query sequence at
+ * the bottom.
+ * ```
+ * ATGG--CGTAGAGC
+ * |||X  |||X|  |
+ * ATGCCCCGTTG--C
+ * ```
+ * In this case, the function seqan3::detail::get_cigar_string will return
+ * the following cigar string when printed: "4M2I5M2D1M". The extended cigar
+ * string would look like this: "3=1X2I3=1X1=2D1=".
+ * \sa seqan3::aligned_sequence_concept
+ */
+template<tuple_like_concept alignment_type>
+//!\cond
+    requires std::tuple_size_v<remove_cvref_t<alignment_type>> == 2
+//!\endcond
+std::string get_cigar_string(alignment_type && alignment,
+                             uint32_t const query_start_pos = 0,
+                             uint32_t const query_end_pos = 0,
+                             bool const extended_cigar = false)
+{
+    return get_cigar_string(std::get<0>(alignment), std::get<1>(alignment),
+                            query_start_pos, query_end_pos, extended_cigar);
 }
 
 } // namespace seqan3::detail

--- a/include/seqan3/io/alignment_file/format_sam.hpp
+++ b/include/seqan3/io/alignment_file/format_sam.hpp
@@ -367,7 +367,8 @@ public:
     }
 
 protected:
-    //!/brief The format version string.
+    //!\privatesection
+    //!\brief The format version string.
     static constexpr char format_version[4] = "1.6";
 
     //!\brief A variable that tracks whether the header_ptr as been written or not.
@@ -381,7 +382,9 @@ protected:
      * \param[in]     field_value The value to print.
      */
     template <typename stream_it_t, typename field_type>
+    //!\cond
         requires std::ranges::ForwardRange<field_type>
+    //!\endcond
     void write_range(stream_it_t & stream_it, field_type && field_value)
     {
         if (ranges::empty(field_value))

--- a/include/seqan3/io/alignment_file/input_options.hpp
+++ b/include/seqan3/io/alignment_file/input_options.hpp
@@ -45,6 +45,7 @@ namespace seqan3
 {
 
 //!\brief The options type defines various option members that influence the behaviour of all or some formats.
+//!\ingroup alignment_file
 struct alignment_file_in_options
 {
     // options to define

--- a/include/seqan3/io/alignment_file/output.hpp
+++ b/include/seqan3/io/alignment_file/output.hpp
@@ -644,6 +644,8 @@ public:
     }
 
 protected:
+    //!\privatesection
+
     //!\brief The file header object.
     std::unique_ptr<alignment_file_header> header_ptr{nullptr};
 

--- a/include/seqan3/io/alignment_file/output_options.hpp
+++ b/include/seqan3/io/alignment_file/output_options.hpp
@@ -45,6 +45,7 @@ namespace seqan3
 {
 
 //!\brief The options type defines various option members that influence the behavior of all or some formats.
+//!\ingroup alignment_file
 struct alignment_file_output_options
 {
     /*!\brief The default plain text line-ending is "\n", but on Windows an

